Source/JavaScriptCore/assembler/MacroAssemblerARM64.h

```diff
@@ -6391,16 +6324,40 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
 
     std::optional<RegisterID> tryFoldBaseAndOffsetPart(BaseIndex address)
     {
-        if (!address.offset)
+        auto immediate = address.offset;
+        if (!immediate)
             return address.base;
-        if (isUInt12(address.offset)) {
-            m_assembler.add<64>(getCachedMemoryTempRegisterIDAndInvalidate(), address.base, UInt12(address.offset));
+
+        if (auto tuple = tryExtractShiftedImm(immediate)) {
+            auto [u12, shift, inverted] = tuple.value();
+            if (!inverted)
+                m_assembler.add<64>(getCachedMemoryTempRegisterIDAndInvalidate(), address.base, u12, shift);
+            else
+                m_assembler.sub<64>(getCachedMemoryTempRegisterIDAndInvalidate(), address.base, u12, shift);
             return memoryTempRegister;
         }
-        if (isUInt12(toTwosComplement(address.offset))) {
-            m_assembler.sub<64>(getCachedMemoryTempRegisterIDAndInvalidate(), address.base, UInt12(toTwosComplement(address.offset)));
-            return memoryTempRegister;
+
+        return std::nullopt;
+    }
+
+    template<typename IntType>
+    std::optional<std::tuple<UInt12, int32_t, bool>> tryExtractShiftedImm(IntType immediate)
+    {
+        if (isUInt12(immediate))
+            return std::tuple { UInt12(immediate), 0, false };
+
+        if (isUInt12(toTwosComplement(immediate)))
```

Comment:
nit: pre-cache the result of `toTwosComplement(immediate)`?

Reply:
Changed.